engine/graphics/src/graphics.cpp

```diff
@@ -929,27 +934,114 @@ namespace dmGraphics
         }
     }
 
+    static void PutShaderResourceBindings(const ShaderDesc::ResourceBinding* bindings, uint32_t bindings_count, dmArray<ShaderResourceBinding>& bindings_out, ShaderResourceBinding::BindingFamily family)
+    {
+        bindings_out.SetCapacity(bindings_count);
+        bindings_out.SetSize(bindings_count);
+
+        for (int i = 0; i < bindings_count; ++i)
+        {
+            ShaderResourceBinding& res = bindings_out[i];
+            res.m_Name                 = strdup(bindings[i].m_Name);
+            res.m_NameHash             = bindings[i].m_NameHash;
+            res.m_Binding              = bindings[i].m_Binding;
+            res.m_Set                  = bindings[i].m_Set;
+            res.m_BlockSize            = bindings[i].m_BlockSize;
+            res.m_Type.m_UseTypeIndex  = bindings[i].m_Type.m_UseTypeIndex;
+            res.m_BindingFamily        = family;
+
+            if (res.m_Type.m_UseTypeIndex)
+                res.m_Type.m_TypeIndex = bindings[i].m_Type.m_Type.m_TypeIndex;
+            else
+                res.m_Type.m_ShaderType = bindings[i].m_Type.m_Type.m_ShaderType;
+        }
+    }
+
+    void CreateShaderMeta(ShaderDesc::Shader* ddf, ShaderMeta* meta)
```

Comment:
Imo we should migrate all the graphics adapters to use this format. I think we could simplify things in the renderer as well for getting uniform locations and attributes and whatnot. We could return the shader meta mapping instead, which contains everything the renderer needs to know.